nslsii/EPS_devices/valves_and_shutters.py

```diff
@@ -0,0 +1,363 @@
+        else:
+            raise nslsiiValves_and_ShuttersValueError(
+                'the kwarg ``stop_str`` in the EpicsTwoStateDevice class'
+                '``__init__`` method needs to be None or needs to match one of'
+                'the kwargs ``state1_str`` or ``state2_str``')
+
+        self._cmd_map = {self._state1_str: self._state1_cmd,
+                         self._state2_str: self._state2_cmd}
+        self._target_map = {self._state1_str: self._state1_val,
+                            self._state2_str: self._state2_val}
+
+        self._time_fmtstr = '%Y-%m-%d %H:%M:%S'
+
+    _state1_cmd = FmtCpt(EpicsSignal,
+                         '{self.prefix}Cmd:{self._state1_pv_uid}-Cmd',
+                         string=True)
+    _state2_cmd = FmtCpt(EpicsSignal,
+                         '{self.prefix}Cmd:{self._state2_pv_uid}-Cmd',
+                         string=True)
+
+    status = Cpt(EpicsSignalRO, 'Pos-Sts', string=True)
+
+    enabled_status = Cpt(EpicsSignalRO, 'Enbl-Sts', string=True)
+
+    _fail_to_state1 = FmtCpt(EpicsSignalRO,
+                             '{self.prefix}Sts:Fail{self._state1_pv_uid}-Sts',
+                             string=True)
+    _fail_to_state2 = FmtCpt(EpicsSignalRO,
+                             '{self.prefix}Sts:Fail{self._state2_pv_uid}-Sts',
+                             string=True)
+
+    def set(self, val):
+        if self._set_st is not None:
+            raise RuntimeError(
+                f'trying to set {self.name} while a set is in progress')
+
+        cmd_sig = self._cmd_map[val]
+        target_val = self._target_map[val]
+
+        st = self._set_st = DeviceStatus(self)
+        enums = self.status.enum_strs
+
+        def state_cb(value, timestamp, **kwargs):
+            value = enums[int(value)]
+            if value == target_val:
+                self._set_st._finished()
+                self._set_st = None
+                self.status.clear_sub(state_cb)
+
+        cmd_enums = cmd_sig.enum_strs
+        count = 0
+
+        def cmd_retry_cb(value, timestamp, **kwargs):
+            nonlocal count
+            value = cmd_enums[int(value)]
+            count += 1
+            if count > self._num_retries:
+                cmd_sig.clear_sub(cmd_retry_cb)
+                st._finished(success=False)
+            if value == 'None':
+                if not st.done:
+                    time.sleep(self._retry_sleep_time)
+                    cmd_sig.set(1)
+                    ts = datetime.datetime.fromtimestamp(
+                        timestamp).strftime(self._time_fmtstr)
+                    print(f'** ({ts}) Had to reactuate {self.name} while'
+                          f'moving to {val}')
+                else:
+                    cmd_sig.clear_sub(cmd_retry_cb)
+
+        cmd_sig.subscribe(cmd_retry_cb, run=False)
+        cmd_sig.set(1)
+        self.status.subscribe(state_cb)
+
+        return st
+
+    def stop(self, success):
```

Comment:
I am not sure that we want to have this as the base-class `stop` logic.  This means that `%mv shutter 'Open'` will 'fail' because it will open and then on the way out the RE will call 'stop' on it which will re-close it. Same comment on 'resume' below.  

This logic is so that when stopping / resuming a scan at XFP we would be sure that the front end shutter would close and stop cooking their sample.  We should check that we do not also need to put this logic in `pause` as I do not recall of the top of my head if RE calls `stop` when suspenders fire (I think it does...).

Reply:
Yes this was an issue that @mrakitin mentioned for existing shutters at CMS as well last week. I have made this change. FYI It does call 'stop' when suspenders fire, as the guys at SIX asked me 2 weeks ago what was the logic for a gate-valve closing when the suspender 'fires'. Once I told them they where actually happy that it did it !-).

I have removed this logic.

Reply:
resolved